libyul/optimiser/DataFlowAnalyzer.cpp

```diff
@@ -345,16 +367,26 @@ void DataFlowAnalyzer::clearKnowledgeIfInvalidated(Expression const& _expr)
 	if (sideEffects.invalidatesStorage())
 		m_storage.clear();
 	if (sideEffects.invalidatesMemory())
+	{
 		m_memory.clear();
+		m_keccak.clear();
+	}
 }
 
 void DataFlowAnalyzer::joinKnowledge(
 	unordered_map<YulString, YulString> const& _olderStorage,
-	unordered_map<YulString, YulString> const& _olderMemory
+	unordered_map<YulString, YulString> const& _olderMemory,
+	map<pair<YulString, YulString>, YulString> const& _olderKeccak
 )
 {
 	joinKnowledgeHelper(m_storage, _olderStorage);
 	joinKnowledgeHelper(m_memory, _olderMemory);
+
+	// TODO is this correct?
+	cxx20::erase_if(m_keccak, mapTuple([&_olderKeccak](auto&& key, auto&& currentValue){
+		YulString const* oldValue = valueOrNullptr(_olderKeccak, key);
+		return !oldValue || *oldValue != currentValue;
+	}));
```

Comment:
So, after `joinKnowledge` `m_keccak` will contain only the keys whose values remained unchanged and are present in both old and new state? Basically, `if` branches and `switch` cases can invalidate parts of state but cannot add anything that would be still useful outside of them?

In case of `if` it sounds about right because there's no `else` in Yul but for `switch` I think it could be more permissive. I.e. if we have the same keccak calculation in all cases (including the default one), we should keep it.

Though it probably does not matter much if even the switch prefix mover (#10742) turned out to be not very useful in practice.